subprojects/docs/src/docs/userguide/dependencyManagementTerminology.adoc

```diff
@@ -0,0 +1,44 @@
+[[dependency_management_terminology]]
+== Dependency Management Terminology
+
+Dependency management comes with a wealth of terminology. Here you can find the most commonly-used terms including references to the user guide to learn about their practical application.
+
+[[sub:terminology_module]]
+Module::
+
+A piece of software that evolves over time e.g. link:https://github.com/google/guava[Google Guava]. Every module has a name. Each release of a module is optimally represented by a <<sub:terminology_version,version>>. For convenient consumption, modules can be hosted in a <<sub:terminology_repository,repository>>.
+
+[[sub:terminology_version]]
+Version::
+
+A <<sub:terminology_module,module>> indicates its revision with the help of a version. Each module version contains a distinct set of changes. In practice there's no limitation to the scheme of the version. Timestamps, numbers, special suffixes like GA are all allowed identifiers. The most widely-used versioning strategy is link:https://semver.org/[semantic versioning]. Semantic versioning is represented by the scheme _major_._minor_._patch_ e.g. `1.4.15`. Each part conveys a special meaning about the backwards-compatibility of the release.
+
+[[sub:terminology_dependency]]
+Dependency::
+
```

Comment:
"a versioned or unversioned module"?
It feels like we haven't cleared in our head what a "dependency" is. I think it's a relationship between 2 modules. But it could also refer to a particular version of a module. But this description seems to indicate that it's both of these things.

A dependency doesn't provide module metadata: a particular release of a module provides module metadata.

I propose that we clarify the term "dependency" to be either:
1. A "dependency requirement": this is part of the metadata of module A v1.0, saying that it has a "dependency requirement" on module B in the range [1.0, 2.0).
2. A "resolved dependency": this would refer to module B v1.4 in the above example. A "resolved dependency" has module metadata, since it is effectively a particular release of a module.

We might use these actual terms, but we _must_ be aware that these are 2 different things when writing documentation.

Reply:
> "a versioned or unversioned module"?

Let me explain my thinking process by example. I hope that at least makes clear how I came up with the definition.

1. A build script can declare a dependency on a versioned module e.g. `commons-lang:commons-lang:1.2.4`.
2. A build script can declare a dependency on another project (which uses its declared version if available) e.g. `project(':abc')`.
3. A build script can declare a dependency on a file that does not have a version e.g. `file('spring.jar')`.

You make good points about the difference between dependency requirement and its metadata. I personally never heard a user use these terms though. Even if we'd mention "dependency requirement" then users will still talk about "a dependency with metadata". Right now that's also reflected in our API by `Dependency`, `ExternalDependency` and `ProjectDependency`.

I believe if we'd go for something like "dependency requirement" then we'd also change that term everywhere we use "dependency" right. I fear that users won't find the right information anymore.

Happy to discuss further. What do others think?

Reply:
> Right now that's also reflected in our API by Dependency, ExternalDependency and ProjectDependency.

These things only refer to the relationship between 2 things, they do _not_ represent the module version that ends up satisfying the dependency. So saying a "dependency with metadata" is wrong, because `Dependency`, `ExternalDependency` and `ProjectDependency` don't have any metadata. They simply define a set of conditions that we use to choose a module version used to choose a module version, and the module version has metadata.

I understand that people (and us) also use the term "dependencies" to refer to the set of jars that are downloaded after we've resolved the "dependencies". But this second use of the term is _not_ represented in our API, AFAIK. Whenever we refer to 'dependency' in our API, we mean the relationship one module version has on another, not the depended-on module version.

Maybe we have to live with this double meaning for the term. But to pretend that these are the same thing will just lead to further confusion.